### ISSUE TITLE
Distogram discrepancies with OpenFold

### ISSUE THREAD

**Issue by amorehead** [lucidrains/alphafold3-pytorch#201]
From Tom in the AlphaFold3 Discord channel:

> The distance binning in alphafold3_pytorch behaves differently from the OpenFold/AF2 implementation (https://github.com/aqlaboratory/openfold/blob/6f63267114435f94ac0604b6d89e82ef45d94484/openfold/utils/feats.py#L93) - where yours seems to pick bin with the closest lower boundary, the AF2 implementation performs actual binning. Is this intentional?

**Comment by vandrw** [comment]
Regarding the binning code, I found two implementations. One in `alphafold3.py` and one in `model_utils.py`. I assume that parts of the code in `alphafold3.py` are being migrated here?

Regarding the issue, I think I see what Tom meant. The OpenFold implementation seems to calculate the linearly distanced bins directly in that function, after which they return a one-hot representation. Based on what I've seen so far in the code, you return a label instead, but the binning is still performed. Please correct me if I'm wrong in this case.

I've made a minimal reproducible example, which showcases the differences between both:
```py
min_bin = 3.25
max_bin = 50.75
num_bins = 39
inf = 1e8
num_samples = 64

input_dists = torch.randn(num_samples) * max_bin + min_bin

# Code from distance_to_bins, alphafold3_pytorch
einx.subtract(
        "... dist, dist_bins -> ... dist dist_bins", input_dists, torch.linspace(min_bin, max_bin, num_bins)
    ).abs().argmin(dim=-1)

# Output
#tensor([38,  8,  0,  5, 31, 38,  0,  0,  0,  0,  0,  0, 20, 38,  0, 28,  0,  0,
#         0, 11, 25, 38, 12,  1,  4,  0,  0,  5, 19,  0,  0, 19, 36, 29, 37, 25,
#        23,  2,  0,  0,  0,  2, 22, 16, 38, 25,  0, 38, 24, 14,  0, 10,  2, 38,
#         0,  0,  0,  0,  0,  0,  0, 38,  0, 38])

# Slightly modified code from dgram_from_positions, openfold
dgram = input_dists.unsqueeze(-1).pow(2)

lower = torch.linspace(min_bin, max_bin, num_bins) ** 2
upper = torch.cat([lower[1:], lower.new_tensor([inf])], dim=-1)
((dgram > lower) * (dgram < upper)).to(dgram.dtype).argmax(dim=-1)

# Output
#tensor([38,  8,  2,  5, 31, 38, 38, 21,  0, 35,  3, 38, 20, 38, 25, 28, 38, 20,
#        38, 11, 25, 38, 11,  0,  4, 24, 10,  5, 18, 38,  0, 18, 35, 29, 37, 25,
#        22,  1, 29,  8, 38,  2, 21, 15, 38, 24,  6, 38, 24, 13,  0,  9,  2, 38,
#        30, 38,  0, 10, 32, 32, 38, 38, 20, 38])
```

Now, the [openfold implementation](https://github.com/aqlaboratory/openfold/blob/6f63267114435f94ac0604b6d89e82ef45d94484/openfold/utils/feats.py#L93-L107) seems to square both the distances and the binning comparisons. If we do the same for your implementation, we'd get the following:
```py
einx.subtract(
        "... dist, dist_bins -> ... dist dist_bins", input_dists.pow(2), torch.linspace(min_bin, max_bin, num_bins).pow(2)
    ).abs().argmin(dim=-1)

#tensor([38,  8,  2,  5, 31, 38, 38, 21,  0, 35,  3, 38, 20, 38, 25, 28, 38, 21,
#        38, 11, 25, 38, 12,  1,  4, 24, 10,  5, 19, 38,  0, 19, 36, 29, 37, 25,
#        23,  2, 29,  9, 38,  2, 22, 16, 38, 25,  6, 38, 24, 14,  0, 10,  2, 38,
#        30, 38,  0, 10, 32, 32, 38, 38, 20, 38])
```

By changing this, it seems to get closer to the output of openfold, but still not the same in some places. I believe that the discrepancy is due to selecting the closest bin during the subtraction step, instead of purely based on a bin's upper and lower values. What do you think, @amorehead? Is this a better approach?

Nevertheless, we would still need to change the code to square the distances and the bins, if we were to assume that their output is correct in this case. While I hope I didn't make any mistakes while trying to port over the openfold code, please review it and let me know if you see anything off!

**Comment by amorehead** [comment]
Hey, @vandrw. Thanks so much for looking into this more closely for us. I lean strongly towards OpenFold's way of calculating distograms, where lower and upper bounds are simply used to determine which bin a distance belongs in. I believe we initially threw this substraction binning logic together early on in the project and never updated it. I think switching to OpenFold's implementation of distogram'ing would require a change in the `distance_to_bins` function's return signature to instead return either their one-hot representation or instead the argmax'd bin labels. Would you be comfortable finding all places in the code where we calculate distograms and refactoring them to use a revised version of OpenFold's distogram logic (giving this function the name `distance_to_dgram` with an optional `return_labels` argument that would instead return the argmax'd one-hot tensors)?

**Comment by vandrw** [comment]
Yes, I'll get started on that now. I'll try to make a commit tonight that fixes this.

Now, regarding the second issue that Tom brought up, I'm not entirely sure where to look. Could you make a separate issue with this and continue there, to keep it organized?

**Comment by lucidrains** [comment]
> Regarding the binning code, I found two implementations. One in `alphafold3.py` and one in `model_utils.py`. I assume that parts of the code in `alphafold3.py` are being migrated here?
> 
> https://github.com/lucidrains/alphafold3-pytorch/blob/496790f539ff7810142ec7a6ec1ebfec01fa0dbc/alphafold3_pytorch/utils/model_utils.py#L40-L53
> 
> The OpenFold implementation seems to calculate the linearly distanced bins directly in that function, after which they return a one-hot representation. Based on what I've seen so far in the code, you return a label instead. Please correct me if I'm wrong in this case.
> 
> I've made a minimal reproducible example, which showcases the differences between both, according to the labels produced:
> 
> ```python
> min_bin = 3.25
> max_bin = 50.75
> num_bins = 39
> inf = 1e8
> num_samples = 64
> 
> distance = torch.randn(num_samples) * max_bin + min_bin
> bins = torch.linspace(min_bin, max_bin, num_bins)
> 
> # Code from distance_to_bins, alphafold3_pytorch
> einx.subtract(
>         "... dist, dist_bins -> ... dist dist_bins", distance, bins
>     ).abs().argmin(dim=-1)
> 
> # Output
> #tensor([38,  8,  0,  5, 31, 38,  0,  0,  0,  0,  0,  0, 20, 38,  0, 28,  0,  0,
> #         0, 11, 25, 38, 12,  1,  4,  0,  0,  5, 19,  0,  0, 19, 36, 29, 37, 25,
> #        23,  2,  0,  0,  0,  2, 22, 16, 38, 25,  0, 38, 24, 14,  0, 10,  2, 38,
> #         0,  0,  0,  0,  0,  0,  0, 38,  0, 38])
> 
> # Slightly modified code from dgram_from_positions, OpenFold
> dgram = distance.unsqueeze(-1).pow(2)
> 
> lower = bins.pow(2) 
> upper = torch.cat([lower[1:], lower.new_tensor([inf])], dim=-1)
> ((dgram > lower) * (dgram < upper)).to(dgram.dtype).argmax(dim=-1)
> 
> # Output
> #tensor([38,  8,  2,  5, 31, 38, 38, 21,  0, 35,  3, 38, 20, 38, 25, 28, 38, 20,
> #        38, 11, 25, 38, 11,  0,  4, 24, 10,  5, 18, 38,  0, 18, 35, 29, 37, 25,
> #        22,  1, 29,  8, 38,  2, 21, 15, 38, 24,  6, 38, 24, 13,  0,  9,  2, 38,
> #        30, 38,  0, 10, 32, 32, 38, 38, 20, 38])
> ```
> 
> Now, the [OpenFold implementation](https://github.com/aqlaboratory/openfold/blob/6f63267114435f94ac0604b6d89e82ef45d94484/openfold/utils/feats.py#L93-L107) seems to square both the distances and the binning comparisons. If we do the same for your implementation, we'd get the following:
> 
> ```python
> einx.subtract(
>         "... dist, dist_bins -> ... dist dist_bins", distance.pow(2), bins.pow(2)
>     ).abs().argmin(dim=-1)
> 
> #tensor([38,  8,  2,  5, 31, 38, 38, 21,  0, 35,  3, 38, 20, 38, 25, 28, 38, 21,
> #        38, 11, 25, 38, 12,  1,  4, 24, 10,  5, 19, 38,  0, 19, 36, 29, 37, 25,
> #        23,  2, 29,  9, 38,  2, 22, 16, 38, 25,  6, 38, 24, 14,  0, 10,  2, 38,
> #        30, 38,  0, 10, 32, 32, 38, 38, 20, 38])
> ```
> 
> By changing this, it seems to get closer to the output of OpenFold, but still not the same in some places. I believe that the discrepancy is due to selecting the closest bin during the subtraction step (as Tom mentioned), instead of purely based on a bin's upper and lower values. What do you think, @amorehead? Is this a better approach?
> 
> Nevertheless, we would still need to change the code to square the distances and the bins, if we were to follow their approach. While I hope I didn't make any mistakes while trying to port over the OpenFold code, please review it and let me know if you see anything off!

ah i see, i assumed the bins were the bin centers, but they are treating adjacent numbers as boundaries, with inf at the right end.

maybe i misunderstand, but in their scheme, where is the bin for `0 - 3.25` though?

**Comment by lucidrains** [comment]
@amorehead @vandrw let's just do it their way, there must be a good reason for the squaring

**Comment by vandrw** [comment]
@lucidrains Understood! I was about to commit as well, but your implementation is much cleaner. Do you mind if I copy it over and add the other changes @amorehead wanted? I will also be importing the new function from `model_utils.py` to avoid having it appear twice.

**Comment by lucidrains** [comment]
@vandrw yes please copy as much as possible!

**Comment by lucidrains** [comment]
should add an assert under debugging flag for bin always needing a 0 on the left

**Comment by vandrw** [comment]
From Tom:
> Thanks for checking this! I've just reviewed and tested the new distance_to_dgram function, it gives the same results as  the OpenFold implementation with the same inputs. One small detail is that https://github.com/lucidrains/alphafold3-pytorch/blob/8e8b7991d6e741cac19d3e36963872d1799cb67d/alphafold3_pytorch/data/template_parsing.py#L151C34-L151C65 I think should have 39 in the call to linspace (so you get 38 actual bins, not counting the last bin for all larger distances)

@amorehead @lucidrains Was there a reason initially for the 38 bins in the following? If we changed the code to what it is now, should we also modify the other number of bins to 65?
https://github.com/lucidrains/alphafold3-pytorch/blob/c026d147a866c45a775d0a927893424bb1d427f5/alphafold3_pytorch/alphafold3.py#L5894-L5896

> I think one could also remove the squaring of distances and bins in your implementation... my naive reading of their code is that they wanted to avoid taking the square root in the distance matrix calculation as it is/was considered an expensive operation in numerical computing?

Replacing the squaring and replacing it by computing the absolute value of the distances returns the same bins in my tests. Should I go ahead and do that as well?

**Comment by lucidrains** [comment]
@vandrw there is no reason, just go ahead and make the PR!

**Comment by amorehead** [comment]
@vandrw, with [this PR](https://github.com/lucidrains/alphafold3-pytorch/pull/213) merged, I think this issue has been resolved. Feel free to voice any concerns you have with the newest implementation or if you think I missed something along the way.
